### ISSUE TITLE
[Critial][Perf] dataframe cache  has significantly perf regression issue when reading json file

### ISSUE THREAD

**Issue by gaoyangxiaozhu** [oap-project/gluten#3456]
### Backend

VL (Velox)

### Bug description

We has a job which run daily to read spark events jsons file to do some simple join op to generate summary metrics.
The job costs increased signficantly from *10 minutes* to now *3 hours* after we sync do last sync of gluten main.

![image](https://github.com/oap-project/gluten/assets/9858245/f9ff2e7e-d7f7-4494-9e5c-9011b46ec526)

The issue only happend when call `dataframe.cache` when reading/processing velox-non supported file format as csv, json, etc.
And it can be easily reproduced with below code

```
spark.read.json(s"${sparkEventsPath}/*/**").cache().collect()
```

before 
![image](https://github.com/oap-project/gluten/assets/9858245/19bf6035-71b5-4b57-a97b-26cddd667797)
![image](https://github.com/oap-project/gluten/assets/9858245/84ce7a57-9362-4d90-8fcd-8e4226ab257b)

after the last week sync:

![image](https://github.com/oap-project/gluten/assets/9858245/f414ef84-9159-482d-8745-000e32251e10)
![image](https://github.com/oap-project/gluten/assets/9858245/ff2b2dcc-ae45-4cb0-80ea-ed93275aabb9)

The issue we found caused due to this PR [[VL] Support columnar table cache](https://github.com/oap-project/gluten/pull/3152)
the PR used for `dafaframe.cache` support but it introduces additional perf issue when cache for csv/json. In shortly, when reading those velox non-supported file,  insteads of do directly convert from internelRow to cachedBatch, cachedBatch to internelRow. It always do additional C2R and R2C convert which 

## Expected behavior

For velox non-support file, please always fallback to use rowbased cachedBatch serializer - DefaultCachedBatchSerializer, insteads of `ColumnarCachedBatchSerializer` , or at least remove addtiaonal C2R and R2C convertion when reading caching for those type file to avoid additional perf issue.

### Spark version

Spark-3.3.x

### Spark configurations

        "spark.sql.broadcastTimeout": "-1",
        "spark.executor.instances": "8",
        "spark.driver.memory": "56g",
        "spark.executor.memory": "20g",
        "spark.executor.cores": "8",
        "spark.driver.cores": "8",
        "spark.sql.sort.enableExtraMetrics": "true",
        "spark.sql.hashAggregate.decimalOverflow.check.enabled": "true",
        "spark.sql.normalize.aggregate.enabled": "true",
        "spark.sql.optimizer.runtime.bloomFilter.maxNumItems": "6000000",
        "spark.sql.adaptive.coalescePartitions.minPartitionSize": "256KB",
        "spark.sql.local.window.optimization.enabled": "true",
        "spark.plugins": "io.glutenproject.GlutenPlugin",
        "spark.gluten.sql.columnar.backend.lib": "velox",
        "spark.memory.offHeap.enabled": "true",
        "spark.memory.offHeap.size": "43g",
        "spark.sql.extensions": "io.glutenproject.GlutenSessionExtensions",
        "spark.executor.memoryOverhead": "1024",
        "spark.shuffle.service.enabled": "false",
        "spark.gluten.sql.columnar.forceshuffledhashjoin": "true",
        "spark.shuffle.manager": "org.apache.spark.shuffle.sort.ColumnarShuffleManager",
        "spark.sql.optimizer.runtime.bloomFilter.applicationSideScanSizeThreshold": "0",
        "spark.gluten.sql.columnar.backend.velox.readerDefaultDirectorySizeGuess":"16384",
        "spark.gluten.sql.columnar.backend.velox.readerDefaultFilePreloadThreshold":"524288",
        "spark.gluten.sql.columnar.backend.velox.hashTableCapacity.scaleUpFactor": "2",
        "spark.sql.optimizer.runtime.bloomFilter.bloomImplClass":"org.apache.spark.util.sketch.BloomFilterImpl",
        "spark.gluten.sql.columnar.maxBatchSize": "10240",
        "spark.gluten.sql.columnar.backend.velox.SplitPreloadPerDriver": "4",
        "spark.sql.files.maxPartitionBytes":"512m",
        "spark.task.cpus": "0.5"

**Comment by gaoyangxiaozhu** [comment]
@zhouyuan @ulysses-you to aware this issue, i think it is critial need fix with high priority.

**Comment by ulysses-you** [comment]
@gaoyangxiaozhu thank you for the report, can you try to tune off columnar table cache `--conf spark.gluten.sql.columnar.tableCache=false` ?

The columnar table cache is a bit hard to auto enable or not. The reason is: we always assume the consume side is offloaded, so even if we add a row to columnar to do the cache, we still have benefit at assume side.

One option is, add a new config to disable columnar cache if the input is row-based.

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu thank you for the report, can you try to tune off columnar table cache `--conf spark.gluten.sql.columnar.tableCache=false` ?
> 
> The columnar table cache is a bit hard to auto enable or not. The reason is: we always assume the consume side is offloaded, so even if we add a row to columnar to do the cache, we still have benefit at assume side.
> 
> One option is, add a new config to disable columnar cache if the input is row-based.

yes, currently we disable the feaure with the flag to unblock our daily run.

**Comment by zhli1142015** [comment]
Should this flag be false by default?

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu thank you for the report, can you try to tune off columnar table cache `--conf spark.gluten.sql.columnar.tableCache=false` ?
> 
> The columnar table cache is a bit hard to auto enable or not. The reason is: we always assume the consume side is offloaded, so even if we add a row to columnar to do the cache, we still have benefit at assume side.
> 
> One option is, add a new config to disable columnar cache if the input is row-based.

em.. just thought even input is row-based, the output may still can be offloaded to columnar batch to leverage velox, so the total time may still has benifit.  Just disable columnar if input is row-based is a option but not good one.

**Comment by zhouyuan** [comment]
CC: @zhztheplayer

**Comment by zhztheplayer** [comment]
> @gaoyangxiaozhu After some attempts, I did not find a good way to resolve this issue. I'm fine if you want to disable it by default.

So it sounds we can disable columnar cache by default until enough optimizations were done?

**Comment by ulysses-you** [comment]
yeah, I think so. @gaoyangxiaozhu do you want to send a pr? thank you!

**Comment by gaoyangxiaozhu** [comment]
> yeah, I think so. @gaoyangxiaozhu do you want to send a pr? thank you!

sure @ulysses-you
